examples/algorithm_progression.py

```diff
@@ -0,0 +1,179 @@
+import argparse
+import concurrent.futures
+import inspect
+import itertools
+import os
+from pathlib import Path
+
+import algorithms
+import numpy as np
+import yaml
+
+import large_image
+
+VARIABLE_LAYERS = ['z', 'c', 't']
+
+
+def create_argparser():
+    argparser = argparse.ArgumentParser(
+        prog='Algorithm Progression',
+        description='Apply an algorithm to an input image '
+                    'for every parameter set in a given parameter space',
+    )
+    argparser.add_argument(
+        'algorithm_code',
+        choices=algorithms.ALGORITHM_CODES.keys(),
+        help='Code to specify which of the available algorithms should be used',
+    )
+    argparser.add_argument('input_filename', help='Path to an image to use as input')
+    argparser.add_argument(
+        '-o',
+        '--output_dir',
+        required=False,
+        help='Name for a new directory in this location, wherein result images will be stored',
+    )
+    argparser.add_argument(
+        '-w',
+        '--num_workers',
+        required=False,
+        default=4,
+        type=int,
+        help='Number of workers to use for processing alogrithm iterations',
+    )
+    argparser.add_argument(
+        '-p',
+        '--param',
+        action='append',
+        required=False,
+        nargs='*',
+        help='A parameter to pass to the algorithm; '
+             'instead of using the default value, '
+             'Pass every item in the number space, '
+             'specified as `--param=param_name,start,end,num_items[,open]`',
+    )
+    return argparser
+
+
+def apply_algorithm(algorithm, input_filename, output_dir, params, param_space):
+    iteration_id = [
+        list(values).index(params[i]) for i, values in enumerate(param_space.values())
+        if len(values) > 1
+    ]
+    filename = f'{algorithm.__name__}_{"_".join([str(v) for v in iteration_id])}.tiff'
+    filepath = Path(output_dir, filename)
+
+    desc = dict(
+        **{'path': filename},
+        **{VARIABLE_LAYERS[i]: v for i, v in enumerate(iteration_id)}
+    )
+
+    source = large_image.open(input_filename)
+    new_source = large_image.new()
+    for tile in source.tileIterator(
+        format=large_image.tilesource.TILE_FORMAT_NUMPY,
+        tile_size=dict(width=2048, height=2048),
+    ):
+        altered_data = algorithm(tile['tile'], *params)
+    new_source.addTile(altered_data, tile['x'], tile['y'])
+    new_source.write(str(filepath), lossy=False)
+    return desc
+
+
+def sweep_algorithm(algorithm, input_filename, input_params, output_dir, max_workers):
+    algorithm_name = algorithm.__name__.replace('_', ' ').title()
+    yaml_dict = {
+        'name': f'{algorithm_name} iterative results',
+        'description': f'{algorithm_name} algorithm performed on {input_filename}',
+        'sources': [{'path': input_filename}],
```

Comment:
If we output multiple axes (or even one axis), the input file doesn't really fit in the matrix of output.  Let's _not_ add it.  If we did add it, it should be on some distinct axis location, as otherwise it will just be overwritten by the output at the 0 values.

Reply:
Ok, I'll take it out.